{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to find the number of unique guide in CRISPRi library\n",
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/milnus/unique_CRISPRi_segments/blob/main/unique_CRISPRi_notebook.ipynb)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To use the notebook it requires a bit of information:\n",
    "1. A file(s) with sequence(s) that are to be searched for guides.\n",
    "2. The sequence that is just after the guide in the plasmid.\n",
    "\n",
    "To give the required inputs: first enter the sequence downstream of primers in the window below. Second, run the notebook. secondly run the code in the window below by pushing the 'play' botton in the window's upper left corner. A required program will install, which takes a couple of minutes and a restart of the underlying 'computer' running the code. When this is done go to `run time` and select then `run all`. in the second code window from the top a: 'Choose files' buttom will appear at the bottom. Using this you can select all your sequencing files. then the script will run and results will be reported at the bottom of the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Install required program\n",
    "# Install conda and seqkit\n",
    "!pip install -q condacolab\n",
    "import condacolab\n",
    "condacolab.install_miniconda()\n",
    "\n",
    "# Install seqkit\n",
    "!conda install bioconda::seqkit -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Input sequence(s)\n",
    "# Import useed packages\n",
    "from subprocess import Popen, PIPE\n",
    "from shlex import split\n",
    "import pandas as pd\n",
    "from google.colab import files\n",
    "\n",
    "# Get inputs\n",
    "#@markdown Change the sequence below to be the one that us just next to the guide sequence\n",
    "sequence_next_to_guides = 'GTTTTAGAGCTAGAAA' #@param {type:\"string\"}\n",
    "\n",
    "#@markdown Set if the sequence above is just up- or down-stream of the guide sequence\n",
    "guide_placement = \"Up-stream\" #@param [\"Up-stream\", \"Down-stream\"]\n",
    "\n",
    "#@markdown Now run this cell to get a 'Choose files' button, where you can select a folder, file, or files to be analysed.\n",
    "uploaded_files = files.upload()\n",
    "\n",
    "# Remove white-space from input sequence after guide\n",
    "sequence_next_to_guides = \"\".join(sequence_next_to_guides.split())\n",
    "\n",
    "# Use of a forward and reverse primer??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title #### Checking input with seqkit\n",
    "### Check that input sequence is valid using seqkit\n",
    "\n",
    "# Check the input sequence\n",
    "cmd = ['echo', f'test_fasta\\n{sequence_next_to_guides}', '|', 'seqkit', 'seq', '-t', 'dna']\n",
    "\n",
    "echo_process = Popen(split(f\"echo '>test_fasta\\n{sequence_next_to_guides}'\"), stdout=PIPE)\n",
    "echo_process.wait()\n",
    "seq_validation_return = Popen(split(\"seqkit seq -t dna -v\"), stdin=echo_process.stdout, stderr=PIPE)\n",
    "seq_validation_return.wait()\n",
    "\n",
    "if seq_validation_return.returncode != 0:\n",
    "\tseqkit_error = str(seq_validation_return.stderr.read()).removesuffix(r\"\\n'\").split('seq: ')[1]\n",
    "\traise TypeError(f'''Input sequence downstream of guide guide contains characters that are non-valid in fasta format.\\nOutput from seqkit:\\n{seqkit_error}''')\n",
    "\n",
    "\n",
    "# Check the input folder or files:\n",
    "for file in uploaded_files:\n",
    "  echo_process = Popen(split(f\"echo '{uploaded_files[file].decode('utf-8')}'\"), stdout=PIPE)\n",
    "  echo_process.wait()\n",
    "  seq_validation_return = Popen(split(\"seqkit seq -t dna -v\"), stdin=echo_process.stdout, stderr=PIPE)\n",
    "  seq_validation_return.wait()\n",
    "\n",
    "  if seq_validation_return.returncode != 0:\n",
    "    seqkit_error = str(seq_validation_return.stderr.read()).removesuffix(r\"\\n'\").split('seq: ')[1]\n",
    "    raise TypeError(f'''Input sequence downstream of guide guide contains characters that are non-valid in fasta format.\\nOutput from seqkit:\\n{seqkit_error}''')\n",
    "\n",
    "print(\"All files are validated by Seqkit\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@title Find number of times sequences occur\n",
    "# Output percentage of sequences with guide\n",
    "# Output an interactive table that can be sorted\n",
    "# Include sequences that does not have a hit - this is important for biological understanding\n",
    "\n",
    "# Constuct dictionary with uniquely identified spacer sequences\n",
    "spacer_dict = {}\n",
    "\n",
    "# Determine if the up- or down-stream region is the guide, based on primer direction:\n",
    "if guide_placement == \"Down-stream\":\n",
    "  extract_region = '1:21'\n",
    "else:\n",
    "  extract_region = '-21:-1'\n",
    "\n",
    "\n",
    "for file in uploaded_files:\n",
    "  echo_process = Popen(split(f\"echo '{uploaded_files[file].decode('utf-8')}'\"), stdout=PIPE)\n",
    "  echo_process.wait()\n",
    "\n",
    "  seqkit_amplicon = Popen(split(f\"seqkit amplicon -F {sequence_next_to_guides} -f -r {extract_region}\"), stdin=echo_process.stdout, stdout=PIPE, stderr=PIPE)\n",
    "  seqkit_amplicon.wait()\n",
    "  \n",
    "  seqkit_fx2tab = Popen(split(\"seqkit fx2tab -s\"), stdin=seqkit_amplicon.stdout, stdout=PIPE, stderr=PIPE)\n",
    "  seqkit_fx2tab.wait()\n",
    "\n",
    "  cut_process = Popen(split(\"cut -f2\"), stdin=seqkit_fx2tab.stdout, stdout=PIPE)\n",
    "  cut_process.wait()\n",
    "\n",
    "  spacer_returned = cut_process.stdout.read().strip()\n",
    "\n",
    "  # Check if no spacer is found\n",
    "  if spacer_returned == b'':\n",
    "    spacer_dict[file] = 'No spacer detected'\n",
    "  else:\n",
    "    spacer_dict[file] = spacer_returned\n",
    "\n",
    "\n",
    "# Count spacers\n",
    "spacer_count_dict = {}\n",
    "for file in spacer_dict:\n",
    "  spacer = spacer_dict[file]\n",
    "  if spacer in spacer_count_dict:\n",
    "    spacer_count_dict[spacer] += 1\n",
    "  else:\n",
    "    spacer_count_dict[spacer] = 1\n",
    "\n",
    "spacer_series = pd.Series(spacer_count_dict)\n",
    "\n",
    "# Print informations of interest\n",
    "num_uniq_spacers = len(spacer_series.drop(spacer_series.filter(like='No', axis=0).index))\n",
    "num_no_spacer = spacer_series.filter(like='No', axis=0)\n",
    "print(f\"There are {num_uniq_spacers} unique spacers in supplied files.\")\n",
    "print(f'\\t{num_uniq_spacers/len(uploaded_files)*100} % of spacers are unique.')\n",
    "\n",
    "print(\"\\n\\n----------------------------------\\n\\n\")\n",
    "\n",
    "if len(num_no_spacer):\n",
    "  print(f\"{num_no_spacer.iloc[0]} files have no identified spacers:\")\n",
    "  for file in spacer_dict:\n",
    "    if spacer_dict[file] == 'No spacer detected':\n",
    "      print(f'\\t{file}')\n",
    "else:\n",
    "  print(\"All sequences had an inserted guide\")\n",
    "\n",
    "# Find the spacers that appear multiple times and find the files that contain them\n",
    "multi_occur_spacers = [spacer for spacer in spacer_count_dict if spacer_count_dict[spacer] > 1 and spacer != 'No spacer detected']\n",
    "\n",
    "print(\"\\n\\n----------------------------------\\n\\n\")\n",
    "if (len(multi_occur_spacers)):\n",
    "  for spacer in multi_occur_spacers:\n",
    "    print(f\"Spacer: {spacer.decode('utf-8')} was found multiple times.\\nThe spacer was found in files:\")\n",
    "    for file in spacer_dict:\n",
    "      if spacer_dict[file] == spacer:\n",
    "        print(f\"\\t{file}\")\n",
    "else:\n",
    "  print(\"All spacers occured only once\")\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
